File: src/main/java/xyz/devvydont/smprpg/items/blueprints/resources/mining/TinFamilyBlueprint.java
package xyz.devvydont.smprpg.items.blueprints.resources.mining;

import org.bukkit.inventory.ItemStack;
import xyz.devvydont.smprpg.items.CustomItemType;
import xyz.devvydont.smprpg.items.ItemClassification;
import xyz.devvydont.smprpg.items.base.CustomItemBlueprint;
import xyz.devvydont.smprpg.items.interfaces.ICompressible;
import xyz.devvydont.smprpg.items.interfaces.ISellable;
import xyz.devvydont.smprpg.services.ItemService;
import xyz.devvydont.smprpg.util.extensions.ItemExtensionsKt;

public class TinFamilyBlueprint extends CustomItemBlueprint implements ICompressible, ISellable {

    public TinFamilyBlueprint(ItemService itemService, CustomItemType type) {
        super(itemService, type);
    }

    @Override
    public ItemClassification getItemClassification() {
        return ItemClassification.MATERIAL;
    }

    @Override
    public CompressionStep getDecompressor() {
        return switch (getCustomItemType()) {
            case TIN_BLOCK -> new CompressionStep((ICompressible) itemService.getBlueprint(CustomItemType.TIN_INGOT), 1, 9);
            case ENCHANTED_TIN -> new CompressionStep((ICompressible) itemService.getBlueprint(CustomItemType.TIN_BLOCK), 1, 9);
            case ENCHANTED_TIN_BLOCK -> new CompressionStep((ICompressible) itemService.getBlueprint(CustomItemType.ENCHANTED_TIN), 1, 9);
            case TIN_SINGULARITY -> new CompressionStep((ICompressible) itemService.getBlueprint(CustomItemType.ENCHANTED_TIN_BLOCK), 1, 9);
            default -> null;
        };
    }

    @Override
    public CompressionStep getCompressor() {
        return switch (getCustomItemType()) {
            case TIN_INGOT -> new CompressionStep((ICompressible) itemService.getBlueprint(CustomItemType.TIN_BLOCK), 9, 1);
            case TIN_BLOCK -> new CompressionStep((ICompressible) itemService.getBlueprint(CustomItemType.ENCHANTED_TIN), 9, 1);
            case ENCHANTED_TIN -> new CompressionStep((ICompressible) itemService.getBlueprint(CustomItemType.ENCHANTED_TIN_BLOCK), 9, 1);
            case ENCHANTED_TIN_BLOCK -> new CompressionStep((ICompressible) itemService.getBlueprint(CustomItemType.TIN_SINGULARITY), 9, 1);
            default -> null;
        };
    }

    @Override
    public int getWorth(ItemStack itemStack) {
        return ItemExtensionsKt.calculateCompressedWorth(this, itemStack);
    }
}

File: src/main/java/xyz/devvydont/smprpg/items/blueprints/resources/mining/TinFamilyBlueprint.kt
package xyz.devvydont.smprpg.items.blueprints.resources.mining

import org.bukkit.inventory.ItemStack
import xyz.devvydont.smprpg.items.CustomItemType
import xyz.devvydont.smprpg.items.ItemClassification
import xyz.devvydont.smprpg.items.base.CustomItemBlueprint
import xyz.devvydont.smprpg.items.blueprints.craftengine.CraftEngineCompressibleBlueprint
import xyz.devvydont.smprpg.items.interfaces.ICompressible
import xyz.devvydont.smprpg.items.interfaces.ICompressible.CompressionStep
import xyz.devvydont.smprpg.items.interfaces.ISellable
import xyz.devvydont.smprpg.services.ItemService
import xyz.devvydont.smprpg.util.extensions.calculateCompressedWorth

class TinFamilyBlueprint(itemService: ItemService, type: CustomItemType) :
    CraftEngineCompressibleBlueprint(itemService, type), ICompressible, ISellable {

    override val itemClassification get() = ItemClassification.MATERIAL

    override val decompressor: CompressionStep? get() = when (customItemType) {
        CustomItemType.TIN_BLOCK          -> CompressionStep(itemService.getBlueprint(CustomItemType.TIN_INGOT) as ICompressible, 1, 9)
        CustomItemType.ENCHANTED_TIN      -> CompressionStep(itemService.getBlueprint(CustomItemType.TIN_BLOCK) as ICompressible, 1, 9)
        CustomItemType.ENCHANTED_TIN_BLOCK-> CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_TIN) as ICompressible, 1, 9)
        CustomItemType.TIN_SINGULARITY    -> CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_TIN_BLOCK) as ICompressible, 1, 9)
        else -> null
    }

    override val compressor: CompressionStep? get() = when (customItemType) {
        CustomItemType.TIN_INGOT          -> CompressionStep(itemService.getBlueprint(CustomItemType.TIN_BLOCK) as ICompressible, 9, 1)
        CustomItemType.TIN_BLOCK          -> CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_TIN) as ICompressible, 9, 1)
        CustomItemType.ENCHANTED_TIN      -> CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_TIN_BLOCK) as ICompressible, 9, 1)
        CustomItemType.ENCHANTED_TIN_BLOCK-> CompressionStep(itemService.getBlueprint(CustomItemType.TIN_SINGULARITY) as ICompressible, 9, 1)
        else -> null
    }

    override fun getWorth(item: ItemStack) = calculateCompressedWorth(item)
}
